### PR TITLE
Deflake TcpProxyManyConnections test

### PR DESCRIPTION
**Pull request by yanavlasov**
Additional Description:
The reason for flakes is that integration tests oversubscribe CPU by a factor of 4 or 5. This causes timing issues, especially for CPU heavy builds (i.e. sanitizers), or builds that do not run on RBE/EngFlow (coverage, arm).

TcpProxyManyConnections fails in two places when starved for CPU. Creating autonomous upstreams (which is less likely) and then exchanging data over the connections (more likely). Stretching timeouts decreases probability of failure significantly, but does not eliminate it.

Splitting it from the TcpProxyIntegrationTest suite deflakes the tests in the TcpProxyIntegrationTest suite completely in my tests.

If the TcpProxyManyConnections continues to fail, the next step is to eliminate the data exchange step, as the only condition it tests is ability to establish more than 1024 connections, it does not actually need to send any data at all.

Also reserved 4 cores for running the TcpProxyManyConnections test. We may consider reserving more cores for all integration tests to avoid oversubscribing CPU too much.

Risk Level: Low
Testing: Unit Test
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Review by phlax — DISMISSED** [review]
i can follow most of the changes - lgtm, thanks @yanavlasov 

not sure if you want to get a more c++ competent stamp

File: test/integration/tcp_proxy_many_connections_test.cc [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  }
+  for (int i = 0; i < num_connections; ++i) {
+    IntegrationTcpClientPtr& tcp_client = clients[i];
+    // The autonomous upstream is an HTTP upstream, so send raw HTTP.
+    // This particular request will result in the upstream sending a response,
+    // and flush-closing due to the 'close_after_response' header.
+    ASSERT_TRUE(tcp_client->write(
+        "GET / HTTP/1.1\r\nHost: foo\r\nclose_after_response: yes\r\ncontent-length: 0\r\n\r\n",
+        false));
+    tcp_client->waitForHalfClose(TestUtility::DefaultTimeout * 20);
```
Comment (RyanTheOptimist) [inline test/integration/tcp_proxy_many_connections_test.cc:69]: nit: Maybe a comment about why we add a factor of 20 here?

**Review by RyanTheOptimist — APPROVED** [review]
Wow, great analysis! Thanks for fixing this.

**Comment by cpakulski** [comment]
My recent PR was also affected by this. I noticed that it failed only in `compiled_time_options` CI run. Was it specific to how CPUs are allocated in that CI type?

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by yanavlasov** [comment]
> My recent PR was also affected by this. I noticed that it failed only in `compiled_time_options` CI run. Was it specific to how CPUs are allocated in that CI type?

It fails in many different builds. Arm, tsan/asan, coverage. Not specific to compiled_time_options

**Pull request merged**